# MNIST using Neural Networks

Run the statement to import the TensorFlow library:

In [18]:
import tensorflow as tf
import numpy as np

Run the statement to load and prepare the [MNIST dataset](http://yann.lecun.com/exdb/mnist/) and to convert the samples from integers to floating-point numbers:

In [2]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11493376/11490434 [==============================] - 6s 0us/step


In [28]:
x_train = np.reshape(x_train,(60000,28,28,1))
x_test = np.reshape(x_test,(10000,28,28,1))

Build the model with a single layer and relu activation function:

In [29]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28,28,1)),
  tf.keras.layers.MaxPooling2D((2, 2)),
  tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
  tf.keras.layers.MaxPooling2D((2, 2)),
  tf.keras.layers.Conv2D(128, (3, 3), activation='relu'),
  tf.keras.layers.MaxPooling2D((2, 2)),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10)
])
# try to add convolutions and change the no of layers and no of nodes/neurons
# to add a new layer
# tf.keras.layers.Dense(10, activation='relu') 
# to add convolutions 
# tf.keras.layers.Conv2D(28, (3, 3), activation='relu', input_shape=(28, 28)) 
# tf.keras.layers.MaxPooling2D((2, 2))

In [30]:
model.summary()

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_18 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_20 (Conv2D)           (None, 3, 3, 128)         73856     
_________________________________________________________________
max_pooling2d_11 (MaxPooling (None, 1, 1, 128)         0         
_________________________________________________________________
flatten_6 (Flatten)          (None, 128)              

Choosing our loss function, in this case crossentropy:

In [31]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [34]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

Adjusting the model parameters to minimize the loss of our model: 

In [35]:
model.fit(x_train, y_train, epochs=5, validation_split = 0.2)

Epoch 1/5
1500/1500 [==============================] - 24s 16ms/step - loss: 0.2127 - accuracy: 0.9340 - val_loss: 0.0897 - val_accuracy: 0.9746
Epoch 2/5
1500/1500 [==============================] - 26s 17ms/step - loss: 0.0704 - accuracy: 0.9787 - val_loss: 0.0662 - val_accuracy: 0.9799
Epoch 3/5
1500/1500 [==============================] - 26s 17ms/step - loss: 0.0519 - accuracy: 0.9834 - val_loss: 0.0639 - val_accuracy: 0.9812
Epoch 4/5
1500/1500 [==============================] - 26s 17ms/step - loss: 0.0387 - accuracy: 0.9883 - val_loss: 0.0511 - val_accuracy: 0.9843
Epoch 5/5
1500/1500 [==============================] - 26s 18ms/step - loss: 0.0308 - accuracy: 0.9901 - val_loss: 0.0566 - val_accuracy: 0.9836


Checking the model's performance on the Validation-set.

In [36]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - loss: 0.0568 - accuracy: 0.9825


[0.05680268630385399, 0.9825000166893005]

Try to improve the accuracy to 97% by changing the number of layers and/or adding convolutions.